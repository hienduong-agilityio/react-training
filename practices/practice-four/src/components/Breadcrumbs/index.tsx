// Libraries
import { Link } from 'react-router-dom';
import { memo } from 'react';

// SVG
import homeIcon from '@public/images/homeIcon.svg';
import arrowRight from '@public/images/arrowRightIcon.svg';

// Constants
import { ROUTE } from '@/constants';

export type BreadcrumbItem = {
  // id: The breadcrumb Items id
  id?: number;
  // label: The label to display for the breadcrumb item
  label?: string;
  // href: Optional URL for the breadcrumb item
  href?: string;
};

interface IBreadcrumbsProps {
  // items: Array of breadcrumb items
  items?: BreadcrumbItem[];
}

/**
 * Breadcrumbs component
 *
 * @returns {JSX.Element} Breadcrumbs element
 */
const Breadcrumbs = ({ items = [{ label: 'Home', href: ROUTE.ROOT }] }: IBreadcrumbsProps): JSX.Element => {
  return (
    <nav aria-label='breadcrumb'>
      <ul className='flex items-center space-x-1 md:space-x-2 rtl:space-x-reverse'>
        <li className='inline-flex items-center'>
          <Link to={ROUTE.ROOT} className='flex items-center text-sm font-medium text-gray-400 hover:text-primary-500'>
            <img src={homeIcon} alt='home' className='w-4 h-4 me-2.5 p-0' />
          </Link>
        </li>
        {items.map((item) => (
          <li key={item.label || item.id} className='inline-flex items-center'>
            <img src={arrowRight} alt='Arrow Right' className='rtl:rotate-180 w-6 h-6 text-gray-400 mx-1' />
            {item.href ? (
              <Link to={item.href} className='ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2'>
                {item.label}
              </Link>
            ) : (
              <span className='ms-1 text-sm font-medium text-gray-400 md:ms-2'>{item.label}</span>
            )}
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default memo(Breadcrumbs);
