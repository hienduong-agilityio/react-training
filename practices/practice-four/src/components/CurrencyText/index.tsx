export interface ICurrencyTextProps extends React.HTMLAttributes<HTMLParagraphElement> {
  // Currency: Optional estimated cost of the project, displayed as a financial estimate.
  currency?: number;
}

/**
 * CurrencyText component
 *
 * @returns {JSX.Element} - CurrencyItem element
 */
export const CurrencyText = ({ currency = 0, className = '' }: ICurrencyTextProps): JSX.Element => {
  return <p className={`${className}`}>{currency ? `US$ ${currency}k` : '-'}</p>;
};
