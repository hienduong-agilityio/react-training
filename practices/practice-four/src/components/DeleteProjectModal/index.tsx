// Components
import Modal from '@/components/common/Modal';
import Button from '@/components/common/Button';
import Spinner from '@/components/common/Spinner';

// Enums
import { BUTTON_COLORS, BUTTON_VARIANTS } from '@/enums/theme';

// Hooks
import { useProject } from '@/hooks/useProject';

// Store
import { ToastStore } from '@/stores/toastStore';

// Types
import type { IProjectItemProps } from '@/interfaces';

export interface IDeleteProjectModalProps {
  // projectId: The project id.
  projectId: string;
  // isModalOpen: A boolean indicating whether the modal is open.
  isModalOpen: boolean;
  // onCloseModal: A function to be called when the modal is requested to be closed.
  onCloseModal: () => void;
}

/**
 * The custom DeleteProjectModal component
 *
 * @returns {JSX.Element} - The DeleteProjectModal element
 */
export const DeleteProjectModal = ({
  projectId,
  isModalOpen = false,
  onCloseModal = () => {}
}: IDeleteProjectModalProps): JSX.Element => {
  const project = useProject({ id: projectId || '' });
  const { showToast } = ToastStore();

  const {
    mutate: { deleteProject },
    data: projectDetail,
    isMutating
  } = project;

  const projectName = (projectDetail as IProjectItemProps)?.projectName || 'this project';

  const handleDeleteProjectSuccess = () => {
    showToast('Project Deleted successfully', 'success');
    onCloseModal();
  };

  const handleDeleteProjectError = () => {
    showToast('Failed to deleted project', 'error');
    onCloseModal();
  };

  const handleDelete = () => {
    if (projectId) {
      deleteProject(projectId, {
        onSuccess: handleDeleteProjectSuccess,
        onError: handleDeleteProjectError
      });
    }
  };

  return (
    <Modal isOpen={isModalOpen} onClose={onCloseModal}>
      <div className='bg-light rounded-t-xl p-4 w-[440px]'>
        <p className='text-lg font-semibold text-gray-900 mb-3'>Delete Project</p>
        <p className='leading-6 text-gray-600'>
          Are you sure you want to delete <strong>{projectName}</strong>? If you delete, it will be permanently lost.
        </p>
      </div>
      <div className='bg-light rounded-b-xl flex justify-end gap-4 border-t-2 py-3'>
        <Button onClick={onCloseModal} variant={BUTTON_VARIANTS.OUTLINED}>
          Cancel
        </Button>
        <Button
          onClick={handleDelete}
          disabled={isMutating}
          variant={BUTTON_VARIANTS.CONTAINED}
          color={BUTTON_COLORS.DANGER}
          customClasses='mr-5'
        >
          {isMutating ? <Spinner /> : 'Delete'}
        </Button>
      </div>
    </Modal>
  );
};
