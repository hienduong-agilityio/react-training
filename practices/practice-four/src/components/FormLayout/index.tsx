interface FormLayoutProps {
  title: string;
  subtitle?: string;
  children: React.ReactNode;
}

export const FormLayout = ({ title, subtitle, children }: FormLayoutProps) => (
  <section>
    <div className='flex rounded-t-xl items-center bg-white justify-between p-4'>
      <p className='text-xl font-semibold text-gray-900'>{title}</p>
    </div>
    {subtitle && <p className='text-gray-600 mt-1 mb-4'>{subtitle}</p>}
    {children}
  </section>
);
