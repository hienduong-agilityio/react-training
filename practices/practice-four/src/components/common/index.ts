export { default as Avatar } from './Avatar';
export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Modal } from './Modal';
export { default as Pagination } from './Pagination';
export { default as SelectField } from './SelectField';
export * from './SkeletonLoader';
export { default as Spinner } from './Spinner';
export { default as StatusLabel } from './StatusLabel';
export { default as Tag } from './Tag';
export { default as Toast } from './ToastMessage';
