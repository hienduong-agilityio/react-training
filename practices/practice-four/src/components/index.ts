export { default as Breadcrumbs } from './Breadcrumbs';
export * from './CurrencyText';
export * from './DeleteProjectModal';
export * from './ErrorBoundary';
export * from './ErrorDisplay';
export * from './ProjectDetailSkeleton';
export { default as ProjectFilterDropdown } from './ProjectFilterDropdown';
export * from './ProjectFilterDropdown';
export { default as ProjectItem } from './ProjectItem';
export * from './ProjectItem';
export * from './ProjectOptionsDropdown';
export * from './ProjectTable';
export { default as ResourceTag } from './ResourcesTag';
export { default as SearchBox } from './SearchBox';
export * from './SuspenseLoader';
export { default as Timeline } from './Timeline';
export * from './ToastContainer';
export * from './FormLayout';
export * from './ProjectFormFields';
export * from './SubmitForm';

// Common
export * from './common';
