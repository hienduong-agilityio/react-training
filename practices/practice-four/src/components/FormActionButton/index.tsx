// Libraries
import { useFormStatus } from 'react-dom';

// Components
import { Button, Spinner } from '@/components';

// Enums
import { BUTTON_COLORS, BUTTON_SIZES, BUTTON_VARIANTS } from '@/enums';

interface IFormActionButtonProps {
  children: React.ReactNode;
  type?: 'submit' | 'button';
  color?: BUTTON_COLORS;
  variant?: BUTTON_VARIANTS;
  size?: BUTTON_SIZES;
  disabledOnPending?: boolean;
  showSpinner?: boolean;
  onClick?: () => void;
  customClass?: string;
}

export const FormActionButton = ({
  children,
  type = 'submit',
  variant = BUTTON_VARIANTS.CONTAINED,
  size = BUTTON_SIZES.LARGE,
  color = BUTTON_COLORS.PRIMARY,
  disabledOnPending = true,
  showSpinner = true,
  customClass = '',
  onClick
}: IFormActionButtonProps) => {
  const { pending } = useFormStatus();

  return (
    <Button
      type={type}
      disabled={disabledOnPending && pending}
      variant={variant}
      size={size}
      color={color}
      onClick={onClick}
      customClasses={customClass}
    >
      {pending && showSpinner ? <Spinner /> : children}
    </Button>
  );
};
