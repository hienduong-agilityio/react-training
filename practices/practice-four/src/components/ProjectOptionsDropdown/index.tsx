// Libraries
import { useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

// Components
import { Button } from '@/components';

// Hooks
import { useClickOutside } from '@/hooks';

// SVG
import addIcon from '@public/images/menuIcon.svg';

// Constants
import { ROUTE } from '@/constants';

export interface IProjectOptionsDropdown {
  // projectID: The ID of the project to be managed
  projectId: string;
  // onDeleteProject: The function to open delete modal
  onDeleteProject: (projectId: string) => void;
}

/**
 * The custom ProjectOptionsDropdown component
 *
 * @returns {JSX.Element} - The ProjectOptionsDropdown element
 */
export const ProjectOptionsDropdown = ({
  projectId = '',
  onDeleteProject = () => {}
}: IProjectOptionsDropdown): JSX.Element => {
  const location = useLocation();

  const [isOptionOpen, setIsOptionOpen] = useState(false);

  const handleClose = () => setIsOptionOpen(false);
  const attachRef = useClickOutside(handleClose);

  const toggleDropdown = () => setIsOptionOpen(!isOptionOpen);

  const handleDeleteClick = () => {
    onDeleteProject(projectId);
    setIsOptionOpen(false);
  };

  return (
    <div className='relative inline-block text-left group' ref={attachRef}>
      <Button onClick={toggleDropdown} customClasses={'hover:border-none hover:bg-gray-100'}>
        <img src={addIcon} alt='Menu' />
      </Button>

      {isOptionOpen && (
        <div className='absolute right-0 z-10 mt-2 bg-white border w-max border-gray-300 rounded-md shadow-lg'>
          <ul>
            <Link to={`${ROUTE.PROJECT}/edit/${projectId}`} state={{ from: location.pathname + location.search }}>
              <li className='px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 cursor-pointer'>Edit</li>
            </Link>
            <li className='px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 cursor-not-allowed'>Send mail</li>
            <li className='px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 cursor-not-allowed'>Details</li>
            <li className='px-4 py-2 text-sm hover:bg-gray-100 cursor-not-allowed text-warning-400'>Archive</li>
            <button
              className='px-4 py-2 text-sm hover:bg-gray-100 cursor-pointer text-danger-400 text-left w-full'
              onClick={handleDeleteClick}
            >
              Delete
            </button>
          </ul>
        </div>
      )}
    </div>
  );
};
