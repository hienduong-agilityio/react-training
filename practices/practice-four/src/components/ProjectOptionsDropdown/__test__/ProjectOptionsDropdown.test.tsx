// Libraries
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

// Components
import { ProjectOptionsDropdown } from '@/components';

const defaultProps = {
  projectId: '123456789',
  onDeleteProject: jest.fn()
};

describe('ProjectOptionsDropdown Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = (props = defaultProps) => {
    return render(
      <MemoryRouter>
        <ProjectOptionsDropdown {...props} />
      </MemoryRouter>
    );
  };

  it('should match snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render the dropdown button', () => {
    renderComponent();
    const buttonElement = screen.getByAltText('Menu');

    expect(buttonElement).toBeInTheDocument();
    fireEvent.click(buttonElement);

    expect(screen.getByText('Edit')).toBeInTheDocument();
    expect(screen.getByText('Send mail')).toBeInTheDocument();
    expect(screen.getByText('Details')).toBeInTheDocument();
    expect(screen.getByText('Archive')).toBeInTheDocument();
    expect(screen.getByText('Delete')).toBeInTheDocument();
  });

  it('should close dropdown when clicking outside', () => {
    const { container } = renderComponent();
    const buttonElement = screen.getByAltText('Menu');

    fireEvent.click(buttonElement);
    expect(screen.getByText('Edit')).toBeInTheDocument();
    fireEvent.mouseDown(container);

    expect(screen.queryByText('Edit')).not.toBeInTheDocument();
  });

  it('should call onDelete when Delete option is clicked', () => {
    renderComponent();
    const buttonElement = screen.getByAltText('Menu');
    fireEvent.click(buttonElement);

    const deleteOption = screen.getByText('Delete');
    fireEvent.click(deleteOption);

    expect(defaultProps.onDeleteProject).toHaveBeenCalledTimes(1);
  });

  it('should not fail when projectId is an empty string', () => {
    renderComponent({ projectId: '', onDeleteProject: jest.fn() });

    const buttonElement = screen.getByAltText('Menu');
    fireEvent.click(buttonElement);

    expect(screen.getByText('Edit')).toBeInTheDocument();
  });
});
