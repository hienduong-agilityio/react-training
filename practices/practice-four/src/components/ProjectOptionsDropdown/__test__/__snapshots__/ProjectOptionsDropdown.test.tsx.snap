// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectOptionsDropdown Component should match snapshot 1`] = `
<div>
  <div
    class="relative inline-block text-left group"
  >
    <button
      class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 border-white text-primary-400 hover:border-2 hover:border-primary-100 px-4 py-2 text-base text-black hover:border-none hover:bg-gray-100"
    >
      <img
        alt="Menu"
      />
    </button>
  </div>
</div>
`;
